lib/open_project/translations/models/locales_updater.rb

```diff
@@ -0,0 +1,146 @@
+# todo those ought to be in the gemspec
+require 'tempfile'
+require 'crowdin-api'
+require 'zip'
+require 'pathname'
+require 'fileutils'
+
+require_relative '../helpers/tmp_directory'
+require_relative './git_repository'
+require_relative './i18n_provider'
+require_relative './locales_updater_configuration'
+
+ENGLISH_TRANSLATION_FILE = 'en.yml'
+ENGLISH_JS_TRANSLATION_FILE = 'js-en.yml'
+ACCEPTANCE_LEVEL = ENV['ACCEPTANCE_LEVEL'] || 100
+
+class LocalesUpdater
+  extend TmpDirectory
+
+  def self.update_all_locales_of_all_repos(debug: true)
+    repos_to_update = plugins_with_locales
+
+    within_tmp_directory(delete_if_exists: true, debug: debug) do
+      repos_to_update.each do |plugin_name, specifics|
+        # todo each branch
+        create_i18n_handle(specifics)
+
+        within_tmp_directory(path: File.join(FileUtils.pwd, plugin_name), debug: debug) do
+          git_repo = setup_plugin_repo(specifics[:uri], FileUtils.pwd)
+          git_repo.within_repo do
+            upload_english
+            request_build
+            download_and_replace_locales
+          end
+          commit_and_push_plugin_repo(git_repo, debug)
+        end
+      end
+    end
+  end
+
+  def self.plugins_with_locales
+    configuration[:plugins]
+  end
+
+  def self.configuration
+    LocalesUpdaterConfiguration.configuration
+  end
+
+  def self.branch
+    configuration[:branch]
+  end
+
+  def self.previous_branch
+    configuration[:previous_branch]
+  end
+
+  def self.setup_plugin_repo(uri, path)
+    plugin_repo = GitRepository.new(uri, path)
+    plugin_repo.clone
+
+    plugin_repo.checkout(branch)
+    plugin_repo.merge(previous_branch, strategy: :ours) if previous_branch
+    plugin_repo
+  end
+
+  def self.commit_and_push_plugin_repo(plugin_repo, debug)
+    plugin_repo.add('config/locales')
+    plugin_repo.commit('update locales from crowdin')
+    plugin_repo.push(branch) unless debug
+  end
+
+  def self.create_i18n_handle(configuration_hash)
+    @i18n_provider ||= begin
+      project_id = configuration_hash[:project_id]
+      api_key = configuration_hash[:api_key]
+      version = configuration_hash[:version]
+      I18nProvider.new(project_id, api_key, version)
+    end
+  end
+
+  def self.upload_english
+    # either add or update the english (js) translation file
+    titles = {
+      ENGLISH_TRANSLATION_FILE => 'OpenProject Wording',
+      ENGLISH_JS_TRANSLATION_FILE => 'OpenProject JavaScript Wording'
+    }
+    [ENGLISH_TRANSLATION_FILE, ENGLISH_JS_TRANSLATION_FILE].each do |translation_file|
+      path_to_translation = File.join 'config', 'locales', translation_file
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.